iceoryx_dds/source/iceoryx_dds/dds/fastdds_data_reader.cpp

```diff
@@ -0,0 +1,287 @@
+}
+
+iox::cxx::expected<iox::dds::DataReaderError> iox::dds::FastDDSDataReader::takeNext(
+        uint8_t* const buffer,
+        const uint64_t& bufferSize)
+{
+    // validation checks
+    if (!m_isConnected.load())
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::NOT_CONNECTED);
+    }
+    if (buffer == nullptr)
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_RECV_BUFFER);
+    }
+
+    // take next sample and copy into buffer
+    Mempool::Chunk data{static_cast<uint32_t>(bufferSize), buffer};
+    eprosima::fastdds::dds::SampleInfo nextInfo;
+    eprosima::fastrtps::types::ReturnCode_t ret = m_dataReader->take_next_sample(&data, &nextInfo);
+    if (ret == eprosima::fastrtps::types::ReturnCode_t::RETCODE_NO_DATA)
+    {
+        // no samples available
+        return iox::cxx::success<>();
+    }
+    else if (ret != eprosima::fastrtps::types::ReturnCode_t::RETCODE_OK)
+    {
+        LogError() << m_logName << "take_next_sample returned and error";
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_DATA);
+    }
+
+    // valid size
+    auto sampleSize = data.size;
+    if (sampleSize == 0)
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_DATA);
+    }
+    if (bufferSize < sampleSize)
+    {
+        // provided buffer is too small.
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::RECV_BUFFER_TOO_SMALL);
+    }
+
+    // copy data into the provided buffer
+    auto bytes = data.payload;
+    std::copy(bytes, bytes + sampleSize, const_cast<uint8_t*>(buffer));
+
+    return iox::cxx::success<>();
+}
+
+iox::cxx::expected<uint64_t, iox::dds::DataReaderError> iox::dds::FastDDSDataReader::take(
+        uint8_t* const buffer,
+        const uint64_t& bufferSize,
+        const iox::cxx::optional<uint64_t>& maxSamples)
+{
+    if (!m_isConnected.load())
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::NOT_CONNECTED);
+    }
+    if (buffer == nullptr)
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_RECV_BUFFER);
+    }
+
+    // get size of the sample
+    auto peekResult = peekNextSize();
+    if (peekResult.has_value())
+    {
+        uint64_t sampleSize = peekResult.value();
+        if (sampleSize == 0)
+        {
+            return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_DATA);
+        }
+        if (bufferSize < sampleSize)
+        {
+            // Provided buffer is too small.
+            return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::RECV_BUFFER_TOO_SMALL);
+        }
+
+        // take maximum amount possible up to the cap: maxSamples
+        auto bufferCapacity = bufferSize / sampleSize;
```

Comment:
I think this should be `auto bufferCapacity = bufferSize / sampleSize + 1;` otherwise the result may adjusted downwards.
For instance when `bufferSize = 3` and `sampleSize = 2` you would have a `bufferCapacity` of `1` which is maybe not sufficient.